src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
+                /// being removed with each item already in the collection. The dictionary is kept in sync with the <see cref="_listBuilder"/>
+                /// as long as practical. If an operation would result in too much of such work, the dictionary is simply dropped and recreated
+                /// later if/when needed.
+                /// </remarks>
+                public void RemoveMatchingItems(ItemSpec<P, I> itemSpec)
+                {
+                    HashSet<I> items = null;
+                    var dictionaryBuilder = GetOrCreateDictionaryBuilder();
+                    foreach (var fragment in itemSpec.Fragments)
+                    {
+                        IEnumerable<string> referencedItems = fragment.GetReferencedItems();
+                        if (referencedItems != null)
+                        {
+                            // The fragment can enumerate its referenced items, we can do dictionary lookups.
+                            foreach (var spec in referencedItems)
+                            {
+                                string key = FileUtilities.NormalizePathForComparisonNoThrow(spec, fragment.ProjectDirectory);
+                                if (dictionaryBuilder.TryGetValue(key, out var multiValue))
+                                {
+                                    items ??= new HashSet<I>();
+                                    foreach (I item in multiValue)
+                                    {
+                                        items.Add(item);
+                                    }
+                                    dictionaryBuilder.Remove(key);
+                                }
+                            }
+                        }
+                        else
+                        {
+                            // The fragment cannot enumerate its referenced items. Iterate over the dictionary and test each item.
+                            List<string> keysToRemove = null;
+                            foreach (var kvp in dictionaryBuilder)
+                            {
+                                if (fragment.IsMatchNormalized(kvp.Key))
+                                {
+                                    items ??= new HashSet<I>();
+                                    foreach (I item in kvp.Value)
+                                    {
+                                        items.Add(item);
+                                    }
+                                    keysToRemove ??= new List<string>();
+                                    keysToRemove.Add(kvp.Key);
+                                }
+                            }
+
+                            if (keysToRemove != null)
+                            {
+                                foreach (string key in keysToRemove)
+                                {
+                                    dictionaryBuilder.Remove(key);
```

Comment:
ImmutableDictionary has bulk removal support In order to minimize tree mutation (node alloc for rebalance):
https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutabledictionary-2.builder.removerange?view=net-5.0

Reply:
Thank you, I've fixed the code to use `RemoveRange`.